lib/common/strings.c

```diff
@@ -901,6 +902,41 @@ crm_strdup_printf(char const *format, ...)
     return string;
 }
 
+/*!
+ * \internal
+ * \brief Safely format a string, asserting on encoding errors.
+ *
+ * This function wraps vsnprintf and checks its return value.
+ * If vsnprintf returns a negative value (indicating an encoding error),
+ * this function will call pcmk__assert to halt execution in debug builds
+ * or log a critical error in release builds.
+ *
+ * \param[out] str     Buffer to store the formatted string.
+ * \param[in]  size    Size of the buffer.
+ * \param[in]  format  Format string.
+ * \param[in]  ...     Variable arguments for the format string.
+ *
+ * \return Number of characters that would have been written if \p size was
+ *         sufficiently large (not including the terminating null byte),
+ *         or a negative value if an encoding error occurred (though this
+ *         function asserts in that case).
+ * \note The behavior is otherwise identical to snprintf.
+ */
+int
+pcmk__snprintf(char *str, size_t size, const char *format, ...)
```

Comment:
Optionally we could use the `restrict` keywords in the ways that the `snprintf()` function signature does